File: ABC133/E.py
import sys
sys.setrecursionlimit(10**7)

MOD = 1000000007
N, K = map(int, input().split())
edges = [[] for _ in range(N)]
for i in range(N-1):
    a, b = map(lambda x: int(x)-1, input().split())
    edges[a].append(b)
    edges[b].append(a)

colors = [K] + [0 for _ in range(N-1)]


def dfs(parent, edge, n):
    for child in edges[edge]:
        if child != parent:
            colors[child] = max(0, n)
            n -= 1
            dfs(edge, child, K-2)


def solve():
    dfs(0, 0, K-1)
    
    ans = K
    for color in colors[1:]:
        ans *= color
        ans %= MOD
    print(ans)


if __name__ == '__main__':
    solve()

File: ABC133/E.nim
import strutils, sequtils, math

var
    N, K: int
    graph: seq[seq[int]]
    colors: seq[int]
const MOD = 10^9 + 7


proc dfs(cur, parent, n_color: int) =
    var n = n_color
    for child in graph[cur]:
        if child != parent:
            colors[child] = max(0, n)
            n -= 1
            dfs(child, cur, K - 2)


proc solve() =
    (N, K) = stdin.readLine.split.map(parseInt)
    graph = newSeqWith(N + 1, newSeq[int]())
    var a, b: int
    for _ in 0..<N - 1:
        (a, b) = stdin.readLine.split.map(parseInt)
        graph[a].add(b)
        graph[b].add(a)
    colors = newSeq[int](N + 1)
    
    dfs(1, 0, K - 1)
    
    var ans = K
    for color in colors[2..N]:
        ans = ans * color mod MOD
    echo ans


when is_main_module:
    solve()
